js_modules/dagit/packages/core/src/runs/QueryfulRunTimeline.tsx

```diff
@@ -0,0 +1,241 @@
+          status: run.status,
+          startTime: run.startTime * 1000,
+          endTime: run.endTime ? run.endTime * 1000 : now,
+        },
+      ];
+    });
+
+    const jobs = [];
+    if (workspaceOrError && workspaceOrError.__typename === 'Workspace') {
+      for (const locationEntry of workspaceOrError.locationEntries) {
+        if (
+          locationEntry.__typename === 'WorkspaceLocationEntry' &&
+          locationEntry.locationOrLoadError?.__typename === 'RepositoryLocation'
+        ) {
+          for (const repository of locationEntry.locationOrLoadError.repositories) {
+            const repoAddress = buildRepoAddress(
+              repository.name,
+              locationEntry.locationOrLoadError.name,
+            );
+            for (const pipeline of repository.pipelines) {
+              const jobKey = makeJobKey(repoAddress, pipeline.name);
+              if (!(jobKey in jobsByKey)) {
+                continue;
+              }
+
+              const schedules = (repository.schedules || []).filter(
+                (schedule) => schedule.pipelineName === pipeline.name,
+              );
+
+              const jobTicks: TimelineRun[] = [];
+              for (const schedule of schedules) {
+                if (schedule.scheduleState.status === InstigationStatus.RUNNING) {
+                  schedule.futureTicks.results.forEach(({timestamp}) => {
+                    const startTime = timestamp * 1000;
+                    if (overlap({start, end}, {start: startTime, end: startTime})) {
+                      jobTicks.push({
+                        id: `${schedule.pipelineName}-future-run-${timestamp}`,
+                        status: 'SCHEDULED',
+                        startTime,
+                        endTime: startTime + 10 * 1000,
+                      });
+                    }
+                  });
+                }
+              }
+
+              const jobRuns = runsByJob[pipeline.name] || [];
+              if (jobTicks.length || jobRuns.length) {
+                jobs.push({
+                  key: jobKey,
+                  jobName: pipeline.name,
+                  repoAddress,
+                  path: workspacePipelinePath({
+                    repoName: repoAddress.name,
+                    repoLocation: repoAddress.location,
+                    pipelineName: pipeline.name,
+                    isJob: pipeline.isJob,
+                  }),
+                  runs: [...jobRuns, ...jobTicks],
+                });
+              }
+            }
+          }
+        }
+      }
+    }
+
+    const earliest = jobs.reduce((accum, job) => {
+      const startTimes = job.runs.map((job) => job.startTime);
+      return {...accum, [job.key]: Math.min(...startTimes)};
+    }, {} as {[jobKey: string]: number});
+
+    setJobsWithRuns(jobs.sort((a, b) => earliest[a.key] - earliest[b.key]));
+    setLoadedWindow(hourWindow);
+  }, [data, start, end, loading, jobsByKey, hourWindow]);
```

Comment:
This looks good to me, but technically this is more of a useMemo than a useEffect because these setters are only called here. I think making it a useEffect just means there's one extra render after the params change and before the state is updated (so it doesn't matter much)...

Reply:
I think you're right -- this is just copied from existing code, so I didn't do much with this, but we can come back and fix it up a bit more.